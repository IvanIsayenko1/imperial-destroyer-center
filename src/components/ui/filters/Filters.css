.filters {
  display: flex;
  gap: 1rem;
  margin-bottom: 2rem;
}

.filters-search,
.filters-sort,
.filters-direction {
  padding: 0.5rem 1rem;
  background: rgba(255, 59, 59, 0.1);
  border: 1px solid rgba(255, 59, 59, 0.2);
  color: var(--rebel-base);
  border-radius: 4px;
  font-size: 0.875rem;
  cursor: pointer;
}

.filters-search {
  flex: 1;
  cursor: text;
}

.filters-sort-container {
  display: flex;
  gap: 0.5rem;
  align-items: center;
}

.filters-sort {
  min-width: 150px;
}

.filters-direction {
  min-width: 120px;
}

.filters-clear {
  padding: 0.5rem 1rem;
  background: rgba(255, 59, 59, 0.2);
  border: 1px solid rgba(255, 59, 59, 0.3);
  color: var(--rebel-base);
  border-radius: 4px;
  font-size: 0.875rem;
  cursor: pointer;
  transition: all 0.2s ease;
  white-space: nowrap;
}

.filters-clear:hover {
  background: rgba(255, 59, 59, 0.3);
  box-shadow: 0 0 10px rgba(255, 59, 59, 0.2);
}

@media (max-width: 768px) {
  .filters {
    flex-direction: column;
  }

  .filters-sort-container {
    flex-wrap: wrap;
  }

  .filters-sort,
  .filters-direction {
    flex: 1;
    min-width: unset;
  }

  .filters-clear {
    width: 100%;
  }
}
