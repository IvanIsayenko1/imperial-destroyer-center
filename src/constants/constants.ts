// pagination
export const ELEMENTS_PER_PAGE = 10;

// SWAPI
export const FETCH_LIMIT = 100; // Limit for fetching data from SWAPI
export const BASE_URL = "https://www.swapi.tech/api/";
export const PLANETS_URL = `${BASE_URL}planets/`;
export const STARSHIPS_URL = `${BASE_URL}starships/`;
export const VEHICLES_URL = `${BASE_URL}vehicles/`;

// sort
export enum SORT_DIRECTIONS {
  ASC = "asc",
  DESC = "desc",
}
