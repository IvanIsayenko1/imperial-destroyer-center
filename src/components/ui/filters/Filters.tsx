import React from "react";
import "./Filters.css";
import { SORT_DIRECTIONS } from "@constants/constants";

interface FilterState {
  searchTerm: string;
  sortBy: string;
  sortDirection: SORT_DIRECTIONS;
}

interface FilterCallbacks {
  onSearchChange: (value: string) => void;
  onSortChange: (value: string) => void;
  onSortDirectionChange: (direction: SORT_DIRECTIONS) => void;
  onClearFilters: () => void;
}

interface FiltersProps {
  state: FilterState;
  callbacks: FilterCallbacks;
  config: {
    inputPlaceholder: string;
    sortOptions: { label: string; value: string }[];
  };
}

export const Filters: React.FC<FiltersProps> = ({
  state: { searchTerm, sortBy, sortDirection },
  callbacks: {
    onSearchChange,
    onSortChange,
    onSortDirectionChange,
    onClearFilters,
  },
  config: { inputPlaceholder, sortOptions },
}) => {
  const hasActiveFilters = searchTerm || sortBy;

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === "Escape") {
      onClearFilters();
    }
  };

  return (
    <div
      className="filters"
      onKeyDown={handleKeyDown}
      data-testid="filters-container"
    >
      <input
        type="text"
        placeholder={inputPlaceholder}
        value={searchTerm}
        onChange={(e) => onSearchChange(e.target.value)}
        className="filters-search"
      />

      <div className="filters-sort-container">
        <select
          value={sortBy}
          onChange={(e) => onSortChange(e.target.value)}
          className="filters-sort"
          data-testid="filters-sort"
        >
          <option value="">Sort by...</option>
          {sortOptions.map((option) => (
            <option value={option.value} key={option.value}>
              {option.label}
            </option>
          ))}
        </select>

        {sortBy && (
          <select
            value={sortDirection}
            onChange={(e) =>
              onSortDirectionChange(e.target.value as SORT_DIRECTIONS)
            }
            className="filters-direction"
            data-testid="filters-direction"
          >
            <option value={SORT_DIRECTIONS.ASC}>Ascending</option>
            <option value={SORT_DIRECTIONS.DESC}>Descending</option>
          </select>
        )}

        {hasActiveFilters && (
          <button onClick={onClearFilters} className="filters-clear">
            Clear Filters
          </button>
        )}
      </div>
    </div>
  );
};
