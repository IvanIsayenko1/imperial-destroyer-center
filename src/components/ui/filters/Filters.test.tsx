import { render, screen, fireEvent } from "@testing-library/react";
import { Filters } from "./Filters";
import { SORT_DIRECTIONS } from "@constants/constants";

describe("Filters component", () => {
  const mockOnSearchChange = vi.fn();
  const mockOnSortChange = vi.fn();
  const mockOnSortDirectionChange = vi.fn();
  const mockOnClearFilters = vi.fn();

  const sortOptions = [
    { label: "Option 1", value: "option1" },
    { label: "Option 2", value: "option2" },
  ];

  const defaultProps = {
    state: {
      searchTerm: "",
      sortBy: "",
      sortDirection: SORT_DIRECTIONS.ASC,
    },
    callbacks: {
      onSearchChange: mockOnSearchChange,
      onSortChange: mockOnSortChange,
      onSortDirectionChange: mockOnSortDirectionChange,
      onClearFilters: mockOnClearFilters,
    },
    config: {
      inputPlaceholder: "Search...",
      sortOptions,
    },
  };

  it("should render input field with correct placeholder", () => {
    render(<Filters {...defaultProps} />);
    expect(screen.getByPlaceholderText("Search...")).toBeInTheDocument();
  });

  it("should handle search term change", () => {
    render(<Filters {...defaultProps} />);
    const input = screen.getByPlaceholderText("Search...") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "New search" } });
    expect(mockOnSearchChange).toHaveBeenCalledWith("New search");
  });

  it("should render sort options", () => {
    render(<Filters {...defaultProps} />);
    const sortBySelect = screen.getByTestId("filters-sort");
    fireEvent.change(sortBySelect, { target: { value: "option1" } });
    expect(mockOnSortChange).toHaveBeenCalledWith("option1");
  });

  it("should render direction select when sortBy is selected", () => {
    render(
      <Filters
        {...defaultProps}
        state={{ ...defaultProps.state, sortBy: "option1" }}
      />
    );
    const directionSelect = screen.getByTestId("filters-direction");
    expect(directionSelect).toBeInTheDocument();
  });

  it("should not render direction select when sortBy is empty", () => {
    render(<Filters {...defaultProps} />);
    const directionSelect = screen.queryByTestId("filters-direction");
    expect(directionSelect).not.toBeInTheDocument();
  });

  it("should handle sort direction change", () => {
    render(
      <Filters
        {...defaultProps}
        state={{ ...defaultProps.state, sortBy: "option1" }}
      />
    );
    const directionSelect = screen.getByTestId("filters-direction");

    fireEvent.change(directionSelect, {
      target: { value: SORT_DIRECTIONS.DESC },
    });

    expect(mockOnSortDirectionChange).toHaveBeenCalledWith(
      SORT_DIRECTIONS.DESC
    );
  });

  it("should render clear filters button when there are active filters", () => {
    render(
      <Filters
        {...defaultProps}
        state={{ ...defaultProps.state, searchTerm: "test" }}
      />
    );
    const clearButton = screen.getByText("Clear Filters");
    expect(clearButton).toBeInTheDocument();
  });

  it("should call onClearFilters when the clear button is clicked", () => {
    render(
      <Filters
        {...defaultProps}
        state={{ ...defaultProps.state, searchTerm: "test" }}
      />
    );
    const clearButton = screen.getByText("Clear Filters");
    fireEvent.click(clearButton);
    expect(mockOnClearFilters).toHaveBeenCalled();
  });

  it("should not render clear filters button when there are no active filters", () => {
    render(<Filters {...defaultProps} />);
    const clearButton = screen.queryByText("Clear Filters");
    expect(clearButton).toBeNull();
  });

  it("should clear filters on Escape key press", () => {
    render(<Filters {...defaultProps} />);

    const filtersContainer = screen.getByTestId("filters-container");
    fireEvent.keyDown(filtersContainer, { key: "Escape" });
    expect(mockOnClearFilters).toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<Filters {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
