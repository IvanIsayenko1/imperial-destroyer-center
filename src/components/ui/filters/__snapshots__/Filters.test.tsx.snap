// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filters component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="filters"
    data-testid="filters-container"
  >
    <input
      class="filters-search"
      placeholder="Search..."
      type="text"
      value=""
    />
    <div
      class="filters-sort-container"
    >
      <select
        class="filters-sort"
        data-testid="filters-sort"
      >
        <option
          value=""
        >
          Sort by...
        </option>
        <option
          value="option1"
        >
          Option 1
        </option>
        <option
          value="option2"
        >
          Option 2
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
